### PR TITLE
Correct examples and README to a new structure from #113

### PR DESCRIPTION
**Pull request by destogl**
- Repair launch file extensions.
- Repair examples.
- Adjust README to a new structure.

### CODE REVIEW

**Review by Karsten1987 — CHANGES_REQUESTED** [review]
I am not in favor of merging this. The launch files are designated to be demos. These should stay as little and as trivial as possible IMO. The diff drive doesn't have to be prefixed, doesn't have to be dynamically configured for other descriptions. I believe the diff drive demo should solely spawn the diffdrive robot we have in the description packages.

But that's just my 2 cents.

**Comment by destogl** [comment]
> I am not in favor of merging this. The launch files are designated to be demos. These should stay as little and as trivial as possible IMO. The diff drive doesn't have to be prefixed, doesn't have to be dynamically configured for other descriptions. I believe the diff drive demo should solely spawn the diffdrive robot we have in the description packages.

I see this opposite. I think we should provide a sensible structure someone can take when integrating their robot. Not only that, but I found it very frequently useful to have parameterizable launch file for the "starting" of the robot. Then this file can be simply reused when implementing a scenario or adding a new one. Especially, when working with students or newbies in ROS I found it easier to provide them sensible structure then copy-pasting many launch files all around and debugging the same errors multiple times.

Still, I understand also your argumentation. Would to make sense to provide booth examples? I would still propose to make the files less "hacky", i.e. to use clean structure of those.

**Review by olivier-stasse — APPROVED** [review]
Thanks a lot @destogl for this PR. If possible there is one suggestion to simplify the code style for the arguments which, I believe is worth and easy to do. But this can be done in a next iteration, this is also true for the launch file simplification.

File: ros2_control_demo_bringup/launch/diffbot_system.launch.py [hunk 1]
```diff
@@ -12,101 +12,150 @@
...
+            default_value="diffbot_base_controller",
+            description="Robot controller to start.",
+        )
+    )
+    declared_arguments.append(
+        DeclareLaunchArgument(
+            "start_rviz",
+            default_value="true",
+            description="Start RViz2 automatically with this launch file.",
+        )
```
[the remark this answers is not in the document]
Comment (destogl) [reply ros2_control_demo_bringup/launch/diffbot_system.launch.py:78]: I agree with you that the "*.launch.py" files are very verbose. Still, I am not certain if this is the right plate to these simplifications/wrappers/helpers.

For what it is worth, would it be better to work with tooling WS on ros2/launch and ros2/launch_ros repositories?

On the other hand, this would slow down this development, so it would maybe make sense to find a common package in the ros2_framework (`controller_manager` maybe?) where we could add those python helpers/libraries. Let's discuss this on the Wednesday's meeting.

**Comment by destogl** [comment]
@Karsten1987, @olivier-stasse can you take another look. I reduced the diffbot file and also added a minimal rrbot file and linked those in the readme.

**Review by Karsten1987 — CHANGES_REQUESTED** [review]
The demos don't work for me:

* diffbot_system.launch.py doesn't start the joint state broadcaster, which results in Rviz flooding the console with TF problems.
* rrbot.launch:
```
 ➭ ros2 launch ros2_control_demo_bringup rrbot.launch.py
[INFO] [launch]: All log files can be found below /Users/karsten/.ros/log/2021-07-06-09-10-20-433563-Karstens-MacBook-Pro.local-54450
[INFO] [launch]: Default logging verbosity is set to INFO
Task exception was never retrieved
future: <Task finished coro=<LaunchService._process_one_event() done, defined at /Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/launch_service.py:274> exception=RuntimeError("Included launch description missing required argument 'description_file' (description: 'URDF/XACRO description file with the robot.'), given: []")>
Traceback (most recent call last):
  File "/Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/launch_service.py", line 276, in _process_one_event
    await self.__process_event(next_event)
  File "/Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/launch_service.py", line 296, in __process_event
    visit_all_entities_and_collect_futures(entity, self.__context))
  File "/Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 45, in visit_all_entities_and_collect_futures
    futures_to_return += visit_all_entities_and_collect_futures(sub_entity, context)
  File "/Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 45, in visit_all_entities_and_collect_futures
    futures_to_return += visit_all_entities_and_collect_futures(sub_entity, context)
  File "/Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/utilities/visit_all_entities_and_collect_futures_impl.py", line 38, in visit_all_entities_and_collect_futures
    sub_entities = entity.visit(context)
  File "/Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/action.py", line 108, in visit
    return self.execute(context)
  File "/Users/karsten/workspace/ros2/ros2_foxy/build/launch/launch/actions/include_launch_description.py", line 148, in execute
    .format(argument.name, argument.description, ', '.join(argument_names))
RuntimeError: Included launch description missing required argument 'description_file' (description: 'URDF/XACRO description file with the robot.'), given: []
```
* rrbot_gazebo doesn't correctly spawn the controllers
```
[spawner.py-5] [INFO] [1625588039.480512657] [spawner_joint_state_broadcaster]: Waiting for /controller_manager services
[spawner.py-5] [INFO] [1625588041.521079775] [spawner_joint_state_broadcaster]: Waiting for /controller_manager services
[spawner.py-5] [INFO] [1625588043.563983759] [spawner_joint_state_broadcaster]: Waiting for /controller_manager services
[spawner.py-5] [INFO] [1625588045.607004251] [spawner_joint_state_broadcaster]: Waiting for /controller_manager services
[spawner.py-5] [ERROR] [1625588047.445145530] [spawner_joint_state_broadcaster]: Controller manager not available
[ERROR] [spawner.py-5]: process has died [pid 55354, exit code 1, cmd '/Users/karsten/workspace/ros2/ros2_control_ws/install/controller_manager/lib/controller_manager/spawner.py joint_state_broadcaster --ros-args'].
```

**Comment by destogl** [comment]
> The demos don't work for me:

Oh, sorry for this... I'll check this one more time...

**Comment by destogl** [comment]
@Karsten1987 they work on my side.

The issue with TF output form rviz is (unfortunatelly) normal because it takes some time until the controller is started. Maybe adding somehow a start-delay would be possible.

For the RRBot examples: please read exactly the README. From there, you can simply copy-paste commands and test the examples.

The gazebo demo is still not working, I didn't got it working and therefore it is not part of the README.

**Comment by Karsten1987** [comment]
the demos on master work for me. The rrbot as well as gazebo as both start the joint state broadcaster correctly. 
Why did you disable the diff drive base controller in the demo?

**Comment by destogl** [comment]
@Karsten1987 I did all tests one more time. Here the report.

## State of current master for me (commit 43a5415d11b0efdf1fd4b79ff536d80b05d631b8):

1. `ros2 launch ros2_control_demo_bringup rrbot_system_position_only.launch.py` - works as expected (`test_forward_position_controller.launch.py` file is not existing, but fun for testing :) - not a big issue)
2. `ros2 launch ros2_control_demo_bringup rrbot_system_position_only_gazebo.launch.py` - does not start controller manager
3. `ros2 launch ros2_control_demo_bringup rrbot_system_multi_interface.launch.py` - cannot be launched
4. `ros2 launch ros2_control_demo_bringup diffbot_system.launch.py` - works as expected

None of the examples starts rviz per default. IMHO is this suboptimal for people having the first contact with the ros2_control...


## State of this PR

1. `ros2 launch ros2_control_demo_bringup rrbot_minimal.launch.py` - works as expected
    `ros2 launch ros2_control_demo_bringup rrbot_system_position_only.launch.py` - works as expected
2. `ros2 launch ros2_control_demo_bringup rrbot_system_position_only_gazebo.launch.py` - does not start controller manager (because it is also up until [testing commit](https://github.com/StoglRobotics-forks/ros2_control_demos/commit/0fb12f518f9ec0ec71dd7513b7c56199df9bac8f) not changed
3. `ros2 launch ros2_control_demo_bringup rrbot_system_multi_interface.launch.py` - works as expected
4. `ros2 launch ros2_control_demo_bringup diffbot_system.launch.py` - works as expected


Today I will add some additional changes to create more "fluid" flow for newcomers and also repair gazebo example. I will ping you again when I have something :)

**Comment by olivier-stasse** [comment]
Dear @destogl  thanks for your PR. 

I tested too the commit `43a5415` and found the same behavior than you apart 
`ros2 launch ros2_control_demo_bringup rrbot_system_position_only_gazebo.launch.py`
where the `controller_manager` is running. Indeed I can see the `joint_state_broadcaster`.

I tested everything by recompiling `gazebo_ros2_control` (956e77063) 
`ros2_control` (7b48dc99) `ros2_controllers` (df5a5faa775) which I believe are the head of the master branches.

As for your pull request, to be honest they are a bit too many different points for my current available bandwidth to test carefully. Would it be possible to break it down in more simpler PR ? 

 * For 1. you are introducing another test file. I am fine with it if you think this is useful for the newcomers.

 * For 2. @Karsten1987 and myself tested it and it is working so it would be nice to understand why it is not working for you. Sorry but I did not understand your remark "(because it is also up until testing commit not changed" could you clarify ?

 * For 3. you are absolutely right, I tested your changes and IMHO it is a necessary change that could/should be merged asap (at least on my setup). Unfortunately for me more changes are probably required: the robot does not display in rviz. Could you confirm @Karsten1987 or @mahaarbo  that the multi_interfaces.py is working for you ?

 * For 4, I did not have time to carefully review the changes.

 * As for the README.md again thanks a lot for this but should not we before that change the documentation to .rst files in the doc directory. We could separate each of the example file. And this would be more coherent with the current state of ros_control. I found that the rst file rendering in github is probably good for the current README.md situation.

 It would be much more easy for me to give feedback if each point could be handled by a different PR. Thanks again for starting this.

**Comment by destogl** [comment]
> It would be much more easy for me to give feedback if each point could be handled by a different PR. Thanks again for starting this.

You are right. I'll split this into multiple PRs coming in during next hour or so.

>     * As for the README.md again thanks a lot for this but should not we before that change the documentation to .rst files in the doc directory. We could separate each of the example file. And this would be more coherent with the current state of ros_control. I found that the rst file rendering in github is probably good for the current README.md situation.

TBH: I find current README, terrible and hard to follow. I would first like to have README at usable state and then I can split this gladly into .rst files and integrate it into our official docs.

> I tested everything by recompiling `gazebo_ros2_control` (956e77063)

This is probably the reason this is not working for me because I am using the released version. If source version has to be used, we should have this noted in the readme.
